src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -28,24 +28,21 @@
 #include "logging/logHandle.hpp"
 #include "runtime/atomic.hpp"
 
-Semaphore AsyncLogWriter::_sem(0);
-Semaphore AsyncLogWriter::_io_sem(1);
-
 class AsyncLogLocker : public StackObj {
- private:
-  static Semaphore _lock;
+  Semaphore* _sem;
+
  public:
-  AsyncLogLocker() {
-    _lock.wait();
+  AsyncLogLocker(Semaphore* sem): _sem(sem) {
+    assert(sem != nullptr && sem->value() == 1,
```

Comment:
The value() check is not necessary. Please do not pollute the Semaphore API with this.

Rather than passing the Semaphore in it should just be accessed directly as there is only one Semaphore to ever be used here. Make AsyncLogLocker a friend class to allow access.

Reply:
Thanks.  I remove Semaphore::value() and declare friend class SemaphoreLocker there. 

My intention is to do a sanity check. A semaphore can be used to implement a low-level lock if and only if its value is 1. It means that only one thread can enter the critical region at any time.  However, it's error-prone for the RAII class to accept any arbitrary semaphore. It's because its value could be any non-negative number.

Reply:
I guess I wasn't clear enough.

Please do not make any changes to the Semaphore class. This "value" is not needed in general and not needed for your sanity checking.

The Semaphore _lock is defined in the AsyncLogWriter class and initialized to 1 so it can be used to mimic a mutual exclusion lock. This does not need to be sanity checked as it is all handled within a small chunk of code. The AsyncLogLocker does not take a semaphore, it is a friend of AsyncLogWriter and accesses _lock directly.

Thanks,
David

Reply:
PS. If it were warranted, the way to do your "sanity check" would be to introduce a BinarySemaphore class that always initializes to a count of 1. But we do not need this.

Reply:
I see. I remove the sanity check.